### ISSUE TITLE
IMPORT INTO very slow for large nodes and unsorted data

### ISSUE THREAD

**Issue by dankinder** [cockroachdb/cockroach#62614]
**Describe the problem**

I'm currently doing a production import of about 100TB of data into 42 nodes (having 960GB SSD each) and it's going much slower than I would expect. The graphs below show that I'm about a week in and I still have only ~10% of the data imported:
![Screen Shot 2021-03-25 at 2 18 49 PM](https://user-images.githubusercontent.com/5198575/112523562-4435d200-8d75-11eb-9c4b-9e1114a2f7b5.png)
![Screen Shot 2021-03-25 at 2 19 10 PM](https://user-images.githubusercontent.com/5198575/112523581-49931c80-8d75-11eb-89eb-b4d30c3b9911.png)
![Screen Shot 2021-03-25 at 2 19 15 PM](https://user-images.githubusercontent.com/5198575/112523593-4bf57680-8d75-11eb-874c-1dbbfb9339f2.png)
![Screen Shot 2021-03-25 at 2 19 22 PM](https://user-images.githubusercontent.com/5198575/112523598-4dbf3a00-8d75-11eb-9a31-dbfeff2ec1d1.png)
![Screen Shot 2021-03-25 at 2 33 01 PM](https://user-images.githubusercontent.com/5198575/112525214-08037100-8d77-11eb-8704-af9ac03645af.png)

Setting `kv.bulk_io_write.concurrent_addsstable_requests` to higher than 1 per [this forum discussion](https://forum.cockroachlabs.com/t/optimal-import-of-very-large-datasets/4167) seemed to speed it up, but it only seems to help to go up to 2 or 3. I did up to 5 and it was a diminishing return.

I increased `kv.bulk_ingest.max_pk_buffer_size` to 512MB, going higher seemed to not provide any more meaningful speedup.

I suspect the issue is how the data is completely unsorted and even pretty random. I noticed a comment from @dt on https://github.com/cockroachdb/cockroach/issues/34809#issuecomment-521284781 that "Import SST ingestion scans existing data and prohibits key shadowing (required for correctness with unique / pk indexes) added ~1% overhead when ingesting non-overlapping SSTs, higher with overlaps." If that's the case it explains why performance continues to get worse the more data there is.

Part of why this dataset is so random is that I had to dedup it before I could put it in CRDB (hence https://github.com/cockroachdb/cockroach/issues/53363), and to do that scalably I had to shard it into 24000 shards based on (hash_of_pk % 24000) then dedup each shard. So as far as data overlapping what already exists in the DB, it's possibly the worst possible situation.

There is a 2nd dataset, even larger, that we want to import into CRDB but the situation there is the same, I'm concerned that it'll be unfeasibly slow.

**Environment:**
Note: we're on v21.1.0-alpha3 because we need [this fix](https://github.com/cockroachdb/cockroach/pull/59204) -- or else we just keep getting S3 connect timeouts. Those are extra bad because a 1 hour import job takes ~24 hours to restore.

 - CockroachDB version: v21.1.0-alpha3 (pebble)
 - Server OS: Centos 7

An extra disclosure on this deployment is that I've got multiple CRDB nodes sharing a host, each host has a 6 disk JBOD, and using the hostname as a label to distribute the data. This is obviously not a recommended setup, but for this issue, the import is even slower if I run 1 node per host with 6 stores. Hence the setup. I have even been tempted to try running multiple CRDB nodes on a single disk, just for this initial import, to get it to scan less data on adding each SSTable. Either that, or work on doing a distributed sort of my data...

Epic CRDB-2340

Jira issue: CRDB-2758

**Comment by dankinder** [comment]
One thing I should clarify: 42 nodes @ 960GB only comes out to 40TB, so importing 100TB would make no sense 😅  but this dataset is extremely compressible, I expect it to be ~4x less in rocksdb/pebble. We will also be expanding the cluster but that hardware is not available yet.

**Comment by dt** [comment]
Interesting. Note that while prefix and snappy compression tend to do very well, we then have replication. But we can worry about that if/when you run out of space -- for now we can try to hasten that, or more ideally the successful completion of the import, a bit.

Few quick questions: 
1) how many CSV files are you passing to the IMPORT command? is there at least one file per node? is it all 24k separate files? 
2) are any nodes CPU-bound? how much memory headroom do we have to play with?
3) What's the schema of the table being imported into? namely PK type and what do secondary indexes look like?

Then I'd probably take a look at the custom charts page under the debug page and pull up `addsstable.delay.total` (with the duration and non-negative rate options). If it is much over zero, I'd also look at `addsstable.delay.enginebackpressure`. If _that_ is non-zero, we want to look at per-node to see which node, and then see if we can tell why that node is slowing us down, which would be if Read Amplification is elevated or pending compaction is elevated.

If we're seeing lots of engine back-pressure due to compactions (or l0 file count/read amp), that'd be really good to know since we did really substantial work in pebble specifically to avoid that during IMPORTs. If that is it, we'd want to dig into the manifest to see what's going on. It could also suggest we'd want to throw even more memory at the pk adder so it can get more sorting done in memory before it throws files to storage.

OTOH, if we _don't_ see elevated `enginebackpressure` explaining our slowness, then we'll want to look at the IMPORT side. If we're not CPU-bound anywhere, and we're not being back pressured by storage and ultimately disk IO, then it seems like under-utilization. That could be if there aren't enough separate files to go around, or could be something else.

**Comment by dankinder** [comment]
Thanks @dt here's what I'm seeing.

1. It's 24k separate files, and I'm importing in in batches that are multiples of the number of nodes.
2. Only a minority of CPU is used, maybe 20-30% (2-3 vCPU per node). That said I have anecdotally observed that often particular nodes are busier than the rest and that shifts around, but the busier nodes are nowhere near CPU-bound.
3. This import is into a single table with no secondary indexes. The table has 22 columns, 6 of them together are the primary key. The PK column values are mostly strings. The 6 columns together are maybe 150 bytes on average but can be up to ~2KB.

Here are the charts you requested, over the same time range as the charts I put in the initial issue:

addsstable.delay.total:
<img width="981" alt="Screen Shot 2021-03-26 at 9 45 40 AM" src="https://user-images.githubusercontent.com/5198575/112640788-0dfd5e80-8e18-11eb-856d-8021d186874e.png">

addsstable.delay.enginebackpressure
<img width="981" alt="Screen Shot 2021-03-26 at 9 46 22 AM" src="https://user-images.githubusercontent.com/5198575/112640862-253c4c00-8e18-11eb-891d-2854f374a550.png">

rocksdb.read-amplification:
<img width="981" alt="Screen Shot 2021-03-26 at 9 47 26 AM" src="https://user-images.githubusercontent.com/5198575/112640965-4ac95580-8e18-11eb-882c-6ddbab93acae.png">

rocksdb.estimated-pending-compaction:
<img width="981" alt="Screen Shot 2021-03-26 at 9 47 50 AM" src="https://user-images.githubusercontent.com/5198575/112641009-59b00800-8e18-11eb-9a64-7812c8e82b8e.png">

If what you're saying is correct then the issue is on the IMPORT side and not due to backpressure...

There are currently 15108 ranges in this database (my current range size is double the default, 256MB-1GB). Would it be worth massively increasing that to reduce the number of ranges and test again?

Also regarding memory: I have more to play with. I just tried `kv.bulk_ingest.max_pk_buffer_size = '4GiB'`, upped from 512MiB where I had it, and increased mem usage some but otherwise did not help. I'm trying 8GiB now just for kicks (I have maybe double that in headroom), but doubt it will help.

**Comment by dankinder** [comment]
Few more notes:

Currently I've had `kv.bulk_io_write.concurrent_addsstable_requests` at 4, if it matters.

Regarding memory, the usual pattern I see is this:
<img width="981" alt="Screen Shot 2021-03-26 at 10 24 33 AM" src="https://user-images.githubusercontent.com/5198575/112646028-7ac72780-8e1d-11eb-9e49-ee2bcc69511e.png">

It peaked at around 4GB even with `kv.bulk_ingest.max_pk_buffer_size = '8GiB'`. That may be because these files of importing have about 4GB of data when decompressed.

**Comment by dankinder** [comment]
Another interesting note: while an import took 4 hours on the table with existing data (about 5TB of data), when I tested created a new table with duplicate schema and importing into that it took 6 minutes. Clearly the "check for duplicate primary key" stuff is causing an exponential decrease in import speed.

**Comment by dt** [comment]
Huh, well that's a very interesting point. We could confirm that is indeed that specific piece of code taking excess time with a quick profile. On the debug page, the link `Profile` (to `/debug/pprof/ui/profile/?seconds=5`) will open a flame graph and I'd guess that one some of the nodes that aren't idle, we'd see a larger block for `pebbleIterator.CheckForKeyCollisions`. I'm still somewhat shocked a the magnitude of the difference though.

**Comment by dankinder** [comment]
I've been playing with things over the past week and here are my findings so far.

When I profiled during an import, only a small portion (maybe 5-10%) of the CPU was being spent on `pebbleIterator.CheckForKeyCollisions`. But nearly 50% was in `runCompaction`, though that varied depending on when I ran the profile.

I improved the import speed by a little over 50% reducing num_replicas from 3 to 1 which I'll put back to 3 at the end. With that change, importing a batch takes ~3h30m on a populated table (now containing about 2TB of data due to reduced replication factor), and ~3m40s on an empty clone of the table, in the same database. So the difference in speed is still massive. This is with `kv.bulk_io_write.concurrent_addsstable_requests = 3` and `kv.bulk_ingest.max_pk_buffer_size = '1 GiB'`.

I still suspect checking for key collisions is the culprit, but it's slow because of disk contention, not CPU. This graph of the past day shows that disk utilization for most nodes is ~10% but for a few it's more like 40%. Those nodes are probably the import bottleneck at any given time.
![Screen Shot 2021-04-05 at 2 56 22 PM](https://user-images.githubusercontent.com/5198575/113613354-1e36ea00-961f-11eb-8968-953df400baad.png)
My theory: ingesting these sstables that span a wide key range is causing a lot of compaction overhead (despite pebble sub-leveling), which causes AddSSTable reads to slow down. Plus, `AddSSTable` is having to read many different disk blocks since the data in the sstable its ingesting is pretty random.

I ran pebble's LVM visualizer to investigate, [here are the results](https://github.com/cockroachdb/cockroach/files/6260188/lsm-visual.html.zip). This cluster is restored from a backup so I think it's only meaningful toward the end. It does look to me like quite a few L0 files are created and L1-L5 often have many of their tables rewritten. But that's as far as I can interpret it.

So the answer on this one may just be that importing data that's highly scattered across the key range is simply going to be slow as the dataset gets larger, due to compaction overhead. If that's true it just means on future imports we'll need to put more work into sorting (or at least, semi-sorting) our data before import.
